### ISSUE TITLE
[SR] Backport: Harden the security where hash values are compared

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#5812]
This is to backport changes from https://www.drupal.org/project/drupal/issues/2400197.

There's at least a small change in password.inc that needs to be backported - not sure if the rest of the changes apply to us.

**Comment by kiamlaluno** [comment]
Mcrypt functions should be avoided since that extension hasn't been maintained for years and it could have security issues nobody would report because the extension has been already removed from PHP. This means it would not even possible to back-port issue fixes.

**Comment by klonos** [comment]
Thanks @indigoxela and @kiamlaluno ...I thought that since we are supporting php 5.6 and up, it might be worth keeping that bit of code around (till we bump the minimum version of php in our system requirements that is). The `function_exists(('mcrypt_create_iv')` in the code should be enough to skip this if the system is using php 7.2+.

Anyway, I'm happy to remove that bit from the PR, but I'll wait for some more feedback before I do that.

**Comment by indigoxela** [comment]
@klonos IMO we shouldn't introduce deprecated php functions in newly added code. @kiamlaluno explained the culprit.

The code you're trying to port has been added in D8.0 and removed in the meantime.

As of php 7.0 random_bytes() is available in php core, so maybe we should fall back to that if it's available and in the rare case when php was compiled without openssl. But isn't that an edge-case, anyway?

**Comment by klonos** [comment]
> The code you're trying to port has been added in D8.1 and removed in the meantime.

Agreed, but only because Drupal dropped support for php5 - we haven't.

> As of php 7.0 random_bytes() is available in php core, so maybe we should fall back to that if it's available...

Right. I've removed the code in question from `bootstrap.inc`, but not for this reason - mainly because a) looking at it now it seems like scope creep here, and b) it seems that there's a plan in D7 to include a polyfill, which we will most likely cross-port. I have created a new issue for that here: #5813

**Comment by avpaderno** [comment]
There are comments about PHP 5.6, but the code used in the PR would work also with PHP 7. (Backdrop dropped the support for PHP 5.6, though. Now, it requires at least PHP 7.1.)

**Comment by avpaderno** [comment]
Apart from adding more code comments and renaming some local variables, the PR essentially replaces `===` with a calls to `hash_equals()`. There are no backward-compatibility issues, since both evaluate to TRUE in the same cases.

**Comment by quicksketch** [comment]
Thanks @avpaderno for reviewing! I went to merge but it had a conflict. I updated the PR, but the changes seem really straight-forward. I'll let testing pass to confirm we're all set.

**Comment by quicksketch** [comment]
Merged https://github.com/backdrop/backdrop/pull/4220 into 1.x and 1.31.x. Thank you @klonos, @indigoxela, and @avpaderno!
